### DISCUSSION TITLE
Assistance with Webhook Extension

### DISCUSSION THREAD

**Discussion by wesleypearson** [alfio-event/alf.io#1258] [Q&A]
Hello! Im attempting to write a general web hook for N8N to allow easy external processing, 

It is showing in the logs its processing but Im not receiving any data.  

Any help would be greatly appreciated. 

![Google Chrome Alf io Admin 14-July-2023 at 12 11 54  1gVSEYYR-1](https://github.com/alfio-event/alf.io/assets/46010020/28ff766b-2969-4eef-a866-47ef5da2ff7d)

```javascript
function getScriptMetadata() {
    return {
        id: 'smsOnCheckIn',
        displayName: 'SMS on Check-In',
        version: 0,
        async: true,
        events: ['TICKET_CHECKED_IN']
    };
}

function cleansePhoneNumber(phoneNumber) {
    // Remove all non-digit characters
    var cleaned = phoneNumber.replace(/\D/g, '');

    // Check if it's an Australian mobile number
    if (cleaned.length === 10 && cleaned.startsWith('04')) {
        // Convert to international format
        cleaned = '+61' + cleaned.substring(2);
    }

    return cleaned;
}

function executeScript(scriptEvent) {
    if (!scriptEvent.ticket) {
        log.warn('No ticket found in script event');
        return;
    }

    var ticket = scriptEvent.ticket;
    
    if (!ticket.additionalInfo || !ticket.additionalInfo.phoneNumber) {
        log.warn('No phone number found for ticket ' + ticket.id);
        return;
    }

    var phoneNumber = cleansePhoneNumber(ticket.additionalInfo.phoneNumber);
    log.info('Sending data to webhook for phone number ' + phoneNumber);

    var message = "Hello, " + ticket.fullName + "! You have been successfully checked in.";

    // Send data to the webhook
    var webhookUrl = 'https://n8n.artefactgroup.au/webhook/cd445cd8-d9d6-4e57-bc95-0537aebebee6';
    var webhookData = {
        ticketId: ticket.id,
        phoneNumber: phoneNumber,
        message: message
    };
    var response = simpleHttpClient.post(webhookUrl, webhookData).asString();

    if (response.status >= 200 && response.status < 300) {
        log.info('Data successfully sent to webhook for phone number ' + phoneNumber);
    } else {
        log.error('Failed to send data to webhook: ' + response.body);
    }
}
```

**Comment by cbellone** [comment]
Hi @wesleypearson,
A couple of suggestions  / remarks:

**1. logging**
the current implementation is logging only on the server log (i.e. `/home/alfio/logs/alfio.log` on your server), so you should check there.
If you want the logs to be visible on the backoffice "extension log" page, you should replace

- `log.info(...)` with `console.log(...)`
- `log.warn(...)` with `console.warn(...)`
- `log.error(...)` with `console.error(...)`

**2. function parameters**

your script is expecting `scriptEvent` to be an object. That's not the case. `scriptEvent` is a String representing the type of event that is being handled by your script in a given moment.
In your case it's always `TICKET_CHECKED_IN`, because your script is subscribed only to that extension event.

Additional objects (like [ticket](https://alf.io/docs/reference/extensions/reference/ticket/#ticket-checked-in)) can be found in the root scope. This means that you can reference the scope variable `ticket` directly, i.e.

```javascript
function executeScript(scriptEvent) {
  if (!ticket.additionalInfo || !ticket.additionalInfo.phoneNumber) {
      log.warn('No phone number found for ticket ' + ticket.id);
      return;
  }
  [...]
}
```

**3. scope object**

as you can see in the documentation, https://alf.io/docs/reference/extensions/reference/ticket/#ticket-checked-in
`ticket` is the only scope variable available for the `TICKET_CHECKED_IN` event, so the following code:

```javascript

  if (!ticket.additionalInfo || !ticket.additionalInfo.phoneNumber) {
      log.warn('No phone number found for ticket ' + ticket.id);
      return;
  }
```

will always print "No phone number found for ticket", because the [Ticket](https://github.com/alfio-event/alf.io/blob/master/src/main/java/alfio/model/Ticket.java) object doesn't have that information.

Please open an issue and we'll try to add `additionalInfo` in the context of the `TICKET_CHECKED_IN` event 

In the meantime, what you can do is:

1. Create another extension for the [TICKET_ASSIGNED](https://alf.io/docs/reference/extensions/reference/ticket/#ticket-assigned) event, where you can access the `additionalInfo` scope variable
2. Persist the `phoneNumber` information somewhere outside of alf.io
3. modify your script to retrieve the `phoneNumber` information from your external storage
